src/openms/source/KERNEL/MSExperiment.cpp

```diff
@@ -214,91 +217,86 @@ namespace OpenMS
 
   @param ms_level MS level to consider for m/z range, RT range, intensity range and ion mobility (if negative, all MS levels are used)
   */
-  void MSExperiment::updateRanges(Int ms_level)
+  void MSExperiment::updateRanges()
   {
     #ifdef OPENMS_ASSERTIONS
-      double rt_min = RangeRT::isEmpty() ? 0 : getMinRT();
-      double rt_max = RangeRT::isEmpty() ? 0 : getMaxRT();
-      double mz_min = RangeMZ::isEmpty() ? 0 : getMinMZ();
-      double mz_max = RangeMZ::isEmpty() ? 0 : getMaxMZ();
-      double int_min = RangeIntensity::isEmpty() ? 0 : getMinIntensity();
-      double int_max = RangeIntensity::isEmpty() ? 0 : getMaxIntensity();
-      double im_min = RangeMobility::isEmpty() ? 0 : getMinMobility();
-      double im_max = RangeMobility::isEmpty() ? 0 : getMaxMobility();
+      double rt_min = combined_ranges_.RangeRT::isEmpty() ? 0 : combined_ranges_.getMinRT();
+      double rt_max = combined_ranges_.RangeRT::isEmpty() ? 0 : combined_ranges_.getMaxRT();
+      double mz_min = combined_ranges_.RangeMZ::isEmpty() ? 0 : combined_ranges_.getMinMZ();
+      double mz_max = combined_ranges_.RangeMZ::isEmpty() ? 0 : combined_ranges_.getMaxMZ();
+      double int_min = combined_ranges_.RangeIntensity::isEmpty() ? 0 : combined_ranges_.getMinIntensity();
+      double int_max = combined_ranges_.RangeIntensity::isEmpty() ? 0 : combined_ranges_.getMaxIntensity();
+      double im_min = combined_ranges_.RangeMobility::isEmpty() ? 0 : combined_ranges_.getMinMobility();
+      double im_max = combined_ranges_.RangeMobility::isEmpty() ? 0 : combined_ranges_.getMaxMobility();
     #endif
 
-    // reset mz/rt/int range
-    this->clearRanges();
+    // Reset all range managers
+    clearRanges();
 
-    // empty
+    // Empty experiment
     if (spectra_.empty() && chromatograms_.empty())
     {
       return;
     }
 
-    // update
+    // Update spectrum ranges
     for (Base::iterator it = spectra_.begin(); it != spectra_.end(); ++it)
-    {
-      if (ms_level < Int(0) || Int(it->getMSLevel()) == ms_level)
-      {
-        // ranges
-        this->extendRT(it->getRT()); // RT
-        // m/z, intensity and ion mobility from spectrum's range
-        it->updateRanges();
-        this->extend(*it);
-      }
-      // for MS level = 1 we extend the range for all the MS2 precursors
-      if (ms_level == 1 && it->getMSLevel() == 2)
+    {      
+      // Update ranges for the spectrum itself
+      it->updateRanges();
+      
+      // Update spectrum range manager with this spectrum's ranges
+      // Add to both general ranges and MS level-specific ranges
+      spectrum_ranges_.extendUnsafe(*it);
+      spectrum_ranges_.extendRT(it->getRT()); // RT is not part of the range of an individual spectrum
+      
+      spectrum_ranges_.extendUnsafe(*it, it->getMSLevel());
+      spectrum_ranges_.extendRT(it->getRT(), it->getMSLevel()); // RT is not part of the range of an individual spectrum
```

Comment:
not sure if I get this right, but you could just do MS levels here and then update the overall range just once using all MS levels

Reply:
right. makes sense.
I think I did not do it because I was unsure about how to handle ms level dependent things like
precusors (technically level - 1).
Do you have an opinion on that?

Reply:
interesting point.
I can only think of one use case, which is TOPPView.
When viewing an MSExp with only MS2 in 2D Mode, you'd want to see to positions of the precursors.
This means (for TV), we would ideally put positions of PC's in the preceding MS-level.
This MS level range then is non-empty, even though there are no actual spectra in the experiment. Not sure if this is confusing.
Especially if you derive the available MS levels from that range information.

Reply:
maybe these 'virtual' ranges, which are not backed up by data should be flagged as 'from_precursor'.

Reply:
how would you flag them without adding rangemanager? already a lot of ranges :)
Maybe we just don't add the precursor information at all und extract this in TOPPView?